src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisVisitor.cs

```diff
@@ -88,6 +104,35 @@ public override MultiValue VisitLiteral (ILiteralOperation literalOperation, Sta
 			return literalOperation.ConstantValue.Value == null ? NullValue.Instance : TopValue;
 		}
 
+		public override MultiValue VisitBinaryOperator (IBinaryOperation operation, StateValue argument)
```

Comment:
I tried this locally and it looks like `operation` has a constant value for many of the tests. Is there a specific case I'm missing? And is it worth checking the ConstantValue anyway for the easy cases?

Reply:
You're right - I was looking at the `ArgumentOperation` which is the parent of this and that one doesn't have a constant value. I'll improve this to use the Roslyns const evaluator when available.

Reply:
You could maybe put it in the top-level `Visit(IOperation)` helper, unless there's some reason we would want to skip that.

Reply:
I'm kind of split on that - it's semi-dangerous. It would mean that the actual specific operation visitor method would not have final say in what the return value is. Maybe not a big deal... I'll probably do it.

Reply:
I went with that approach - Visit(IOperation) now handles constants (if the operation specific visit doesn't provide a value). It solves lot of easy cases and works generally for things like null and strings as well.
I kept the BinaryOperation visitor as there are cases where it still works (if one or both of the operands are local variables for example - that requires data flow to figure out values, so constant detection itself doesn't do it) and also as a sample what we would need to add for other operators if necessary.